### PR TITLE
refactor(frontend): replace react-vis with recharts

### PR DESCRIPTION
**Pull request by jeffspahr**
## Summary
- replace the ROC curve viewer from react-vis to recharts
- pin recharts transitive react-redux to 8.1.3 for React 17 compatibility
- improve ROC tooltip behavior by selecting nearest display point and add regression coverage
- sync the React 17 checklist and remove stale vite migration docs

## Validation
- npm run test:ui -- src/components/viewers/ROCCurve.test.tsx
- npm run lint
- npm run typecheck
- npm run test:server:coverage
- npm run test:ui:coverage (116 files, 1646 tests passed)

### CODE REVIEW

**Comment by jeffspahr** [comment]
Follow-up update from review feedback:

- Added an identity-based cache for ROC chart preprocessing in ROCCurve so buildChartData is not recomputed on every zoom/hover rerender.
- This specifically reduces repeated Map/Set/sort work during drag interactions, while keeping the migration scope small.
- Cache key is configs reference identity (documented in frontend/docs/react-17-upgrade-checklist.md).

Also included in this push:
- semantic reset zoom button (button type="button")
- safer tooltip fallback parsing
- removed unused label{index} chart row fields
- edge-case tests for findNearestDisplayPoint (empty/single point)

Validation rerun:
- npm run test:ui -- src/components/viewers/ROCCurve.test.tsx (14 passed)
- npm run lint -- src/components/viewers/ROCCurve.tsx

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 1]
```diff
@@ -116,130 +146,199 @@ class ROCCurve extends Viewer<ROCCurveProps, ROCCurveState> {
...
-    const baseLineData = Array.from(Array(100).keys()).map(x => ({ x: x / 100, y: x / 100 }));
-
-    const { hoveredValues, lastDrawLocation, highlightIndex } = this.state;
+    const { lastDrawLocation, highlightIndex, refAreaLeft, refAreaRight } = this.state;
+    const chartData = this.buildChartData(datasets);
+    const xDomain = lastDrawLocation
+      ? ([lastDrawLocation.left, lastDrawLocation.right] as [number, number])
+      : ([0, 1] as [number, number]);
+    const colors = this.props.colors || lineColors;
+    const showLegend = this.props.forceLegend || datasets.length > 1;
```
[the remark this answers is not in the document]
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx]: Fixed in commit 294caef0d. Added identity-based memoization for ROC chart preprocessing via `cachedConfigs` and `cachedChartData`, and render now calls `getChartData(configs, datasets)` so zoom/hover rerenders do not rebuild chart data each time.

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 2]
```diff
@@ -116,130 +146,199 @@ class ROCCurve extends Viewer<ROCCurveProps, ROCCurveState> {
...
 
-          {lastDrawLocation && <span>Click to reset zoom</span>}
+          {lastDrawLocation && (
+            <span
+              role='button'
+              onClick={() => this.setState({ lastDrawLocation: null })}
+              style={{ cursor: 'pointer' }}
+            >
+              Click to reset zoom
+            </span>
```
[the remark this answers is not in the document]
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx]: Fixed in commit 294caef0d. The reset control now uses a semantic button element (`<button type="button">`) instead of `span role="button"`, which addresses keyboard/accessibility behavior.

**Comment by jeffspahr** [comment]
CI follow-up: fixed frontend lockfile drift under CI toolchain.

Root cause was npm version skew (`npm ci` passes locally on npm 11 but failed in CI on npm 10.9.3 with `Missing: babel-plugin-macros@3.1.0 from lock file`).

Fix:
- regenerated `frontend/package-lock.json` under Node 22.19.0 / npm 10.9.3
- verified `npm ci` succeeds locally under that same version pair

Commit: 1072ae21a

**Comment by jeffspahr** [comment]
compare-seeded-roc-1280x800
artifact-lineage-from-list-1280x800

Attaching compare screenshots from the 2 screens that use charts.  Note that there are some changes in the line graph, but I got it as close as reasonably possible.

**Comment by srtanish1992** [comment]
@jeffspahr I did some manual testing and found few issues. I have attached screenshots of master vs branch.

**ROC curve under metrics section on Compare runs page.**

1. The grid lines are spaced out due to the intervals on y axis.
2. The text inside the tooltip is not visible.
3. When I hover over the series legend color blue, I see the red color as well. Same happens for blue.
4. The series legend icon is different.
5. The y axis title is not completely visible.

**Lineage explorer graph.**

1. The edges don't connect correctly.


Screenshot 2026-02-17 at 9 32 30 PM
Screenshot 2026-02-17 at 9 35 19 PM
Screenshot 2026-02-17 at 9 37 40 PM
Screenshot 2026-02-17 at 9 37 20 PM

**Comment by jeffspahr** [comment]
Thanks for the review @srtanish1992!  I forgot to push the latest commit that my screenshots came from.  Please take another look, and I'll start looking into the issues you called out as well.

**Comment by srtanish1992** [comment]
> Thanks for the review @srtanish1992! I forgot to push the latest commit that my screenshots came from. Please take another look, and I'll start looking into the issues you called out as well.

It's looking much better now 🎉 . The only minor issue is that the last data point/tick is missing on the x and y axis i.e `1` in this case.

Screenshot 2026-02-17 at 10 59 57 PM

File: frontend/src/components/viewers/ROCCurve.tsx
Comment (srtanish1992) [inline frontend/src/components/viewers/ROCCurve.tsx]: I don't see `css.axis` used anywhere. Should we remove this class?

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 3]
```diff
@@ -115,131 +150,234 @@ class ROCCurve extends Viewer<ROCCurveProps, ROCCurveState> {
...
-            color={color.disabledBg}
+          <CartesianGrid />
+          <XAxis
+            type='number'
+            dataKey='x'
+            domain={xDomain}
+            ticks={this.xTicks}
+            interval={0}
+            tickFormatter={(value: number) => value.toFixed(2)}
+            tick={{ fill: '#666', fontSize: 12, fontWeight: 400 }}
```
Comment (srtanish1992) [inline frontend/src/components/viewers/ROCCurve.tsx]: Should we use tokens from `Css.tsx` file instead of hardcoded values?

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 4]
```diff
@@ -87,19 +115,26 @@ interface ROCCurveProps {
...
   lastDrawLocation: { left: number; right: number } | null;
   highlightIndex: number;
+  refAreaLeft: number | null;
+  refAreaRight: number | null;
 }
 
 class ROCCurve extends Viewer<ROCCurveProps, ROCCurveState> {
+  private cachedConfigs: ROCCurveConfig[] | null = null;
+  private cachedChartData: Array<Record<string, number | null>> = [];
+  private readonly xTicks = Array.from({ length: 20 }, (_, i) => Number((i * 0.05).toFixed(2)));
```
Comment (srtanish1992) [inline frontend/src/components/viewers/ROCCurve.tsx:149]: Changing length to 21 will fix the issue of including `1`.

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 5]
```diff
@@ -87,19 +115,26 @@ interface ROCCurveProps {
...
+  private readonly yTicks = Array.from({ length: 10 }, (_, i) => Number((i * 0.1).toFixed(1)));
```
Comment (srtanish1992) [inline frontend/src/components/viewers/ROCCurve.tsx:150]: Changing length to 11 will fix the issue of including `1`.

File: frontend/src/components/viewers/ROCCurve.tsx [hunk 6]
```diff
@@ -115,131 +150,234 @@ class ROCCurve extends Viewer<ROCCurveProps, ROCCurveState> {
...
-              <div className={css.crosshair}>
-                {hoveredValues.map((value, i) => (
-                  <div key={i} className={css.crosshairLabel}>{`${labels[i]}: ${value.label}`}</div>
-                ))}
-              </div>
-            </Crosshair>
+            <Tooltip
+              cursor={{ stroke: '#9aa0a6' }}
+              content={({ active, payload, label }) => {
+                if (!active || !payload?.length) {
```
Comment (srtanish1992) [inline frontend/src/components/viewers/ROCCurve.tsx]: Just for readability, would recommend extracting the logic for tooltip content into a separate function. And when we upgrade to react 18, we can memoize it so that a new function is not created on each render.

File: frontend/package.json [hunk 7]
```diff
@@ -174,6 +173,9 @@
   },
   "overrides": {
     "nanoid": "3.3.8",
+    "recharts": {
+      "react-redux": "8.1.3"
```
Comment (srtanish1992) [inline frontend/package.json:177]: Do we need to remove this override for react 18 upgrade?

**Comment by srtanish1992** [comment]
@jeffspahr I have completed my code review.

File: frontend/src/components/viewers/ROCCurve.tsx
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx]: Fixed in 9878875d3. Removed the unused `css.axis` style block and dropped the now-unused `fontsize` import from `ROCCurve.tsx`.

File: frontend/src/components/viewers/ROCCurve.tsx (hunk 3 above)
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx]: Addressed in 373440e09.

Regression summary: during the `react-vis` -> `recharts` migration, ROC tooltip rendering moved from `Crosshair` (styled by `react-vis/dist/style.css`) to a custom Recharts tooltip container. We kept the dark tooltip background but did not set an explicit text color, so text inherited a dark page color and became low-contrast.

Fixes in this commit:
- Tokenized ROC axis and cursor styling to use shared `Css.tsx` values (`color.grey`, `color.strong`, `color.weak`, `fontsize.small`) instead of local hardcoded literals.\n- Added shared tooltip tokens in `Css.tsx` (`color.tooltipBg`, `color.tooltipShadow`) and switched ROC tooltip styling to use them.
- Explicitly set tooltip text color to `color.background` for both the tooltip container and labels to restore readable contrast.

File: frontend/src/components/viewers/ROCCurve.tsx (hunk 6 above)
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx]: Addressed in commit 510b01519.

For readability, I extracted tooltip rendering logic out of the inline `Tooltip.content` callback into dedicated helpers:
- `resolveHoveredX(...)`
- `buildTooltipRows(...)`
- `renderTooltipRows(...)`
- `renderTooltipContent(...)`

The `Tooltip` now uses a small adapter (`content={tooltipProps => this.renderTooltipContent(...)}`), so `render()` stays easier to scan while preserving existing behavior.

File: frontend/package.json (hunk 7 above)
Comment (jeffspahr) [reply frontend/package.json:177]: Yes. We should keep this override while frontend is on React 17.

Without it, `recharts@3.7.0` resolves `react-redux@9.x`, whose peer deps are `react ^18 || ^19`, so npm raises `ERESOLVE overriding peer dependency` and we end up in an unsupported React 17 combination.

Plan: remove this override in the React 18 upgrade PR, regenerate lockfile, and verify `npm ls react-redux` resolves recharts to `react-redux@9.x` cleanly under React 18.

**Comment by jeffspahr** [comment]
### ROC curve under metrics section on Compare runs page.

> 1. The grid lines are spaced out due to the intervals on y axis.

Fixed

> 2. The text inside the tooltip is not visible.

Fixed

> 3. When I hover over the series legend color blue, I see the red color as well. Same happens for blue.

Fixed. I couldn't reproduce this in the latest commit so please confirm on your end as well.

> 4. The series legend icon is different.

Good catch. I think the newer legend looks better.  I'll keep it unless you have a strong opinion on it.

> 6. The y axis title is not completely visible.

Fixed. Please confirm on your end.

###  Lineage explorer graph.

> 7. The edges don't connect correctly.

Fixed.  Decided to make it better than the original rather than just match it.
image

File: frontend/src/components/viewers/ROCCurve.tsx (hunk 4 above)
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx:149]: My first pass included the 1, but existing behavior actually ends at .9 and .95.  I've matched that for now.  I started a thread in the pipelines slack channel to see if there are any strong opinions on this.  If folks want 1, then I'll follow up with a small fix.

File: frontend/src/components/viewers/ROCCurve.tsx (hunk 5 above)
Comment (jeffspahr) [reply frontend/src/components/viewers/ROCCurve.tsx:150]: My first pass included the 1, but existing behavior actually ends at .9 and .95.  I've matched that for now.  I started a thread in the pipelines slack channel to see if there are any strong opinions on this.  If folks want 1, then I'll follow up with a small fix.

**Comment by droctothorpe** [comment]
/approve
/lgtm

Thank you both! 🙏
